numba_dpex/core/kernel_interface/arg_pack_unpacker.py

```diff
@@ -0,0 +1,291 @@
+                # Store to the repack map
+                self._repack_list.append(
+                    _NumPyArrayPackerPayload(
+                        usm_mem, orig_val, packed_val, packed
+                    )
+                )
+            elif access_type == "write_only":
+                self._repack_list.append(
+                    _NumPyArrayPackerPayload(
+                        usm_mem, orig_val, packed_val, packed
+                    )
+                )
+            else:
+                utils.copy_from_numpy_to_usm_obj(usm_mem, packed_val)
+                self._repack_list.append(
+                    _NumPyArrayPackerPayload(
+                        usm_mem, orig_val, packed_val, packed
+                    )
+                )
+
+        return self._unpack_array_helper(
+            packed_val.size,
+            packed_val.dtype.itemsize,
+            usm_mem,
+            packed_val.shape,
+            packed_val.strides,
+            packed_val.ndim,
+        )
+
+    def _unpack_argument(self, ty, val, access_specifier):
+        """
+        Unpack a Python object into a ctype value using Numba's
+        type-inference machinery.
+
+        Args:
+            ty: The data types of the kernel argument defined as in instance of
+            numba.types.
+            val: The value of the kernel argument.
+
+        Raises:
+            UnsupportedKernelArgumentError: When the argument is of an
+            unsupported type.
+
+        """
+
+        if isinstance(ty, USMNdArrayType):
+            return self._unpack_usm_array(val)
+        elif isinstance(ty, types.Array):
+            return self._unpack_array(val, access_specifier)
+        elif ty == types.int64:
+            return ctypes.c_longlong(val)
+        elif ty == types.uint64:
+            return ctypes.c_ulonglong(val)
+        elif ty == types.int32:
+            return ctypes.c_int(val)
+        elif ty == types.uint32:
+            return ctypes.c_uint(val)
+        elif ty == types.float64:
+            return ctypes.c_double(val)
+        elif ty == types.float32:
+            return ctypes.c_float(val)
+        elif ty == types.boolean:
+            return ctypes.c_uint8(int(val))
+        elif ty == types.complex64:
+            raise UnsupportedKernelArgumentError(ty, val, self._pyfunc_name)
+        elif ty == types.complex128:
+            raise UnsupportedKernelArgumentError(ty, val, self._pyfunc_name)
+        else:
+            raise UnsupportedKernelArgumentError(ty, val, self._pyfunc_name)
+
+    def _pack_array(self):
+        """
+        Copy device data back to host
+        """
+        for obj in self._repack_list:
+            utils.copy_to_numpy_from_usm_obj(obj._usm_mem, obj._packed_val)
+            if obj._packed:
+                np.copyto(obj._orig_val, obj._packed_val)
+
+    def __init__(
+        self, kernel_name, arg_list, argty_list, access_specifiers_list, queue
```

Comment:
I think we need to use consistent variable naming, like `pyfunc_name` instead of `kernel_name`